tokio/src/sync/mpsc/chan.rs

```diff
@@ -199,6 +209,38 @@ where
     pub(crate) fn try_send(&mut self, value: T) -> Result<(), (T, TrySendError)> {
         self.inner.try_send(value, &mut self.permit)
     }
+
+    pub(crate) async fn closed(&mut self) {
+        use std::future::Future;
+        use std::pin::Pin;
+        use std::task::Poll;
+
+        // In order to avoid a race condition, we first request a notification,
+        // **then** check the current value's version. If a new version exists,
+        // the notification request is dropped. Requesting the notification
+        // requires polling the future once.
+        let notified = self.inner.notify_rx_closed.notified();
+        pin!(notified);
+
+        // Polling this for first time will register the waiter and
+        // return `Pending` or return `Ready` right away. If `Ready`
+        // is returned we are done
+        let aquired_lost_notification =
```

Comment:
This logic should no longer be necessary w/ #2839. A poll should not be ready and is just to establish the waiter.